content/en/blog/_posts/2023-12-20-contextual-logging-in-kubernetes-1-29.md

```diff
@@ -0,0 +1,146 @@
+
+On behalf of the [Structured Logging Working Group](https://github.com/kubernetes/community/blob/master/wg-structured-logging/README.md) 
+and [SIG Instrumentation](https://github.com/kubernetes/community/tree/master/sig-instrumentation#readme), 
+we are pleased to announce that the contextual logging feature
+introduced in Kubernetes v1.24 has now been successfully migrated to
+two components (kube-scheduler and kube-controller-manager)
+as well as some directories. This feature aims to provide more useful logs 
+for better troubleshooting of Kubernetes and to empower developers to enhance Kubernetes.
+
+## What is Contextual logging?
+
+[Contextual logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging)
+is based on the [go-logr](https://github.com/go-logr/logr#a-minimal-logging-api-for-go) API. 
+The key idea is that libraries are passed a logger instance by their caller
+and use that for logging instead of accessing a global logger. 
+The binary decides about the logging implementation, not the libraries. 
+The go-logr API is designed around structured logging and supports attaching
+additional information to a logger.
+
+This enables additional use cases:
+
+- The caller can attach additional information to a logger:
+  - [WithName](<https://pkg.go.dev/github.com/go-logr/logr#Logger.WithName>) adds a "logger" key with the names concatenated by a dot as value
+  - [WithValues](<https://pkg.go.dev/github.com/go-logr/logr#Logger.WithValues>) adds key/value pairs
+
+  When passing this extended logger into a function and a function uses it
+  instead of the global logger, the additional information is then included 
+  in all log entries, without having to modify the code that generates the log entries. 
+  This is useful in highly parallel applications where it can become hard to identify 
+  all log entries for a certain operation because the output from different operations gets interleaved.
+
+- When running unit tests, log output can be associated with the current test.
+  Then when a test fails, only the log output of the failed test gets shown by go test.
+  That output can also be more verbose by default because it will not get shown for successful tests.
+  Tests can be run in parallel without interleaving their output.
+
+One of the design decisions for contextual logging was to allow attaching a logger as value to a `context.Context`.
+Since the logger encapsulates all aspects of the intended logging for the call, 
+it is *part* of the context and not just *using* it. A practical advantage is that many APIs 
+already have a `ctx` parameter or adding one has additional advantages, like being able to
+get rid of `context.TODO()` calls inside the functions.
+
+## How to use it
+
+The contextual logging feature is alpha starting from Kubernetes v1.24,
+so it requires the ContextualLogging feature gate to be enabled.
```

Comment:
```suggestion
so it requires the `ContextualLogging` feature gate to be enabled.
```

Reply:
Extra credit: hyperlink feature gate to https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/